packages/@aws-cdk/aws-stepfunctions/lib/state-machine.ts

```diff
@@ -148,6 +165,64 @@ abstract class StateMachineBase extends Resource implements IStateMachine {
       resourceArns: [this.stateMachineArn],
     });
   }
+
+  /**
+   * Grant the given identity permissions to read results from state
+   * machine.
+   */
+  public grantRead(identity: iam.IGrantable): iam.Grant {
+    iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: ['states:ListExecutions'],
+      resourceArns: [this.stateMachineArn],
+    });
+    iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:DescribeExecution',
+        'states:DescribeStateMachineForExecution',
+        'states:GetExecutionHistory',
+      ],
+      resourceArns: [`${this.executionArn}:*`],
+    });
+    return iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:ListStateMachines',
+        'states:ListActivities',
+        'states:DescribeStateMachine',
+        'states:DescribeActivity',
+      ],
+      resourceArns: ['*'],
+    });
+  }
+
+  /**
+   * Grant the given identity permissions to access task responses of the
+   * state machine.
+   */
+  public grantTaskResponse(identity: iam.IGrantable): iam.Grant {
+    return iam.Grant.addToPrincipal({
+      grantee: identity,
+      actions: [
+        'states:SendTaskSuccess',
+        'states:SendTaskFailure',
+        'states:SendTaskHeartbeat',
+      ],
+      resourceArns: ['*'],
```

Comment:
IAM Policy Simulator says this for all three: "This action does not support resource-level permissions. Policies granting access must specify "*" in the resource element."

Reply:
@kaizen3031593

Reply:
I tested this on a state machine with an activity and SendTaskSuccess/Failure/Heartbeat worked correctly when mapped onto the activity. It did NOT work when mapped onto the state machine itself. This means that the IAM Policy Simulator is not correct.

I also tested on a state machine with a lambda function and SendTaskSuccess/Failure/Heartbeat worked when mapped onto the state machine.

At any rate, it seems like `grantTaskResponse()` will also require a `resourceArn` parameter which makes it very similar to `grant()`. Thus, I think it is best to omit `grantTaskResponse()` entirely and move forward with just the `grant()` API.

Reply:
that seems reasonable. It would probably be useful to include an example snippet in the README to further guide users who are trying to grant these permissions

Reply:
I disagree, I think that given how hard it was to discover how to configure this there is still value in the `grantTaskResponse ()` method, additionally it configures all of the [Activity-Level Permissions](https://docs.aws.amazon.com/step-functions/latest/dg/concept-create-iam-advanced.html#concept-create-iam-advanced-activity) which in itself gives value.

Reply:
My issue is that `grantTaskResponse()` can take in both a stateMachineArn or an activityArn depending on what is sending the Task Token (i.e. lambdas can send Task Tokens but are not an activity). Perhaps a solution could be to do a `grantTaskResponseForStateMachine()` and a `grantTaskResponseForActivity()` method to split it up?

Reply:
I agree with @NetaNir that we should add support for task responses, but I think it's fine to defer it to a future PR. The reason I think introducing the `grant` and `grantRead` APIs are reasonable is because they currently block normal use cases (and also unlock all other use cases).

task response is not universal to any state machine and either requires: 
* a task that will use the callback pattern
* an activity task.

In either case, the grant is slightly different and you can also have a state machine with both. 

I'm still in favour of introducing it at a later time (or introduce grant now in a separate PR). 
The ability to issue a `grant` of any sort unlocks use cases that are currently not possible.

An API to enable permissions (`grant`) is a must. Having a convenience/sugar method for task responses (`grantxYZ`) is a nice to have.

Reply:
@shivlaks and I discussed offline and decided to implement a `grantTaskResponse()` API that takes in both a `principal` and an optional `ResourceArn`. The `ResourceArn` should default to the state machine arn, which is the most obvious use case. When given a `ResourceArn` it will make sure that the arn is valid (i.e. is a state machine arn or an activity arn). 

This should alleviate @NetaNir's concerns as it addresses activity-level permissions for `grantTaskReponse()`. It should also encapsulate both activity-level and state-machine-level granularity in one convenience method.